File: BE/CodeGenCommon/cpu_neutral.py

from IR import opcode_tab as o
from BE.Base import serialize
from BE.Base import ir
from BE.Elf import elf_unit


MEMKIND_TO_SECTION = {
    o.MEM_KIND.RO: "rodata",
    o.MEM_KIND.RW: "data",
    # bss missing
}


def MemCodeGenText(mem: ir.Mem, _mod: ir.Unit):
    yield f"# size {mem.Size()}"
    yield f".mem {mem.name} {mem.alignment} {MEMKIND_TO_SECTION[mem.kind]}"
    for d in mem.datas:
        if isinstance(d, ir.DataBytes):
            yield f"    .data {d.count} {serialize.EscapeCStyle(d.data)}"
        elif isinstance(d, ir.DataAddrFun):
            yield f"    .addr.fun {d.size} {d.fun.name}"
        elif isinstance(d, ir.DataAddrMem):
            yield f"    .addr.mem {d.size} {d.mem.name} 0x{d.offset:x}"
        else:
            assert False

    yield ".endmem"


def MemCodeGenBinary(unit: elf_unit.Unit, mem: ir.Mem, addr_reloc_kind):
    unit.MemStart(mem.name, mem.alignment, MEMKIND_TO_SECTION[mem.kind], False)
    for d in mem.datas:
        if isinstance(d, ir.DataBytes):
            unit.AddData(d.count, d.data)
        elif isinstance(d, ir.DataAddrFun):
            unit.AddFunAddr(addr_reloc_kind, d.size, d.fun.name)
        elif isinstance(d, ir.DataAddrMem):
            unit.AddMemAddr(addr_reloc_kind, d.size, d.mem.name, d.offset)
        else:
            assert False
    unit.MemEnd()


def JtbCodeGenSimpleText(jtb: ir.Jtb, addr_size: int):
    yield f".localmem {jtb.name} {addr_size} rodata"
    for i in range(jtb.size):
        bbl = jtb.bbl_tab.get(i, jtb.def_bbl)
        yield f"    .addr.bbl {addr_size} {bbl.name}"
    yield ".endmem"


def JtbCodeGenSimpleBinary(unit: elf_unit.Unit, jtb: ir.Jtb, addr_size: int, addr_reloc_kind):
    unit.MemStart(jtb.name, addr_size, "rodata", True)
    for i in range(jtb.size):
        bbl = jtb.bbl_tab.get(i, jtb.def_bbl)
        unit.AddBblAddr(addr_reloc_kind, addr_size, bbl.name)
    unit.MemEnd()

File: BE/CodeGenCommon/cpu_neutral.cc
#include "BE/CodeGenCommon/cpu_neutral.h"

#include "BE/Base/serialize.h"
#include "Util/parse.h"

namespace cwerg::code_gen_common {

using namespace cwerg;
using namespace cwerg::base;

std::string_view MemKindToSectionName(MEM_KIND kind) {
  switch (kind) {
    case MEM_KIND::RO:
      return "rodata";
    case MEM_KIND::RW:
      return "data";
    default:
      ASSERT(false, "");
      return "";
  }
}

void JtbCodeGenSimpleText(Jtb jtb, std::ostream* output, int addr_size) {
  std::vector<Bbl> table(JtbSize(jtb), JtbDefBbl(jtb));
  for (Jen jen : JtbJenIter(jtb)) {
    table[JenPos(jen)] = JenBbl(jen);
  }
  *output << ".localmem " << Name(jtb) << " " << addr_size << " rodata\n";
  for (Bbl bbl : table) {
    *output << "    .addr.bbl " << addr_size << " " << Name(bbl) << "\n";
  }
  *output << ".endmem\n";
}

void MemCodeGenText(Mem mem, std::ostream* output) {
  *output << "# size " << MemSize(mem) << "\n"
          << ".mem " << Name(mem) << " " << MemAlignment(mem) << " "
          << MemKindToSectionName(MemKind(mem)) << "\n";
  for (Data data : MemDataIter(mem)) {
    uint32_t size = DataSize(data);
    Handle target = DataTarget(data);
    int32_t extra = DataExtra(data);
    if (Kind(target) == RefKind::STR) {
      size_t len = size;
      char buffer[4096];
      if (len > 0) {
        len = BytesToEscapedString({StrData(Str(target)), len}, buffer);
      }
      buffer[len] = 0;
      *output << "    .data " << extra << " \"" << buffer << "\"\n";
    } else if (Kind(target) == RefKind::FUN) {
      *output << "    .addr.fun " << size << " " << Name(Fun(target)) << "\n";
    } else {
      ASSERT(Kind(target) == RefKind::MEM, "");
      *output << "    .addr.mem " << size << " " << Name(Mem(target))
              << std::hex << " 0x" << extra << std::dec << "\n";
    }
  }

  *output << ".endmem\n";
}



}  // namespace cwerg::code_gen_common